File: apps/api/internal/assistant/context_recall.go
package assistant

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/Ciao1019/Petrichor/apps/api/ent"
	"github.com/Ciao1019/Petrichor/apps/api/ent/assistantmessage"
	"github.com/Ciao1019/Petrichor/apps/api/internal/ai"
	"github.com/Ciao1019/Petrichor/apps/api/internal/vector"
)

const (
	contextRecallTopK       = 4
	contextRecallMinScore   = 0.25
	contextRecallExcerptMax = 500
	contextRecallEmbedBatch = 16
)

type recalledSnippet struct {
	MessageID string
	Score     float64
	Excerpt   string
}

type recallInput struct {
	UserID            int64
	ThreadID          int64
	Query             string
	ExcludeMessageIDs []int64
	Limit             int
}

type recallDeps struct {
	SQL *sql.DB
	Gen *ai.Generation
}

func sanitizeRecallExcerpt(text string) string {
	replacers := []*regexp.Regexp{
		regexp.MustCompile(`sk-[a-zA-Z0-9]{10,}`),
		regexp.MustCompile(`(?i)(api[_-]?key|token|secret|password|cookie)\s*[:=]\s*\S+`),
		regexp.MustCompile(`(?i)bearer\s+[a-zA-Z0-9._\-]+`),
		regexp.MustCompile(`(?i)executionOutcome[\s\S]{0,200}`),
	}
	next := text
	next = replacers[0].ReplaceAllString(next, "[redacted]")
	next = replacers[1].ReplaceAllString(next, "$1=[redacted]")
	next = replacers[2].ReplaceAllString(next, "Bearer [redacted]")
	next = replacers[3].ReplaceAllString(next, "[confirmation omitted]")
	next = regexp.MustCompile(`\s+`).ReplaceAllString(next, " ")
	next = strings.TrimSpace(next)
	if len([]rune(next)) <= contextRecallExcerptMax {
		return next
	}
	return string([]rune(next)[:contextRecallExcerptMax]) + "…"
}

func recallRelevantHistory(ctx context.Context, client *ent.Client, deps recallDeps, input recallInput) []recalledSnippet {
	query := strings.TrimSpace(input.Query)
	if query == "" || input.ThreadID <= 0 {
		return nil
	}
	limit := input.Limit
	if limit <= 0 {
		limit = contextRecallTopK
	}
	if deps.SQL != nil && deps.Gen != nil {
		if snippets := recallByVector(ctx, client, deps, input, limit); len(snippets) > 0 {
			return snippets
		}
	}
	return recallByFTS(ctx, client, input, limit)
}

func recallByVector(ctx context.Context, client *ent.Client, deps recallDeps, input recallInput, limit int) []recalledSnippet {
	ok, err := ai.HasEmbeddingConfig(ctx, client, input.UserID)
	if err != nil || !ok {
		return nil
	}
	go ensureThreadMessageEmbeddingsBestEffort(context.WithoutCancel(ctx), client, deps, input)

	modelCfg, err := ai.ResolveEmbeddingConfig(ctx, client, input.UserID, nil)
	if err != nil {
		return nil
	}
	q := input.Query
	if len([]rune(q)) > 4000 {
		q = string([]rune(q)[:4000])
	}
	vecs, err := deps.Gen.Embed(ctx, modelCfg, []string{q})
	if err != nil || len(vecs) == 0 || len(vecs[0]) == 0 {
		return nil
	}
	literal := vector.FormatPGVector(vecs[0])

	var rows *sql.Rows
	if len(input.ExcludeMessageIDs) > 0 {
		placeholders := make([]string, len(input.ExcludeMessageIDs))
		args := make([]any, 0, 4+len(input.ExcludeMessageIDs))
		args = append(args, input.ThreadID, input.UserID, literal)
		for i, id := range input.ExcludeMessageIDs {
			placeholders[i] = fmt.Sprintf("$%d", i+4)
			args = append(args, id)
		}
		args = append(args, limit)
		limitIdx := len(args)
		q := fmt.Sprintf(`
			select message_id, excerpt_md,
				(1 - (embedding <=> $3::vector))::float8 as score
			from petrichor_assistant_message_embedding
			where thread_id = $1 and user_id = $2
			  and embedding is not null
			  and message_id not in (%s)
			order by embedding <=> $3::vector
			limit $%d
		`, strings.Join(placeholders, ","), limitIdx)
		rows, err = deps.SQL.QueryContext(ctx, q, args...)
	} else {
		rows, err = deps.SQL.QueryContext(ctx, `
			select message_id, excerpt_md,
				(1 - (embedding <=> $3::vector))::float8 as score
			from petrichor_assistant_message_embedding
			where thread_id = $1 and user_id = $2
			  and embedding is not null
			order by embedding <=> $3::vector
			limit $4
		`, input.ThreadID, input.UserID, literal, limit)
	}
	if err != nil {
		return nil
	}
	defer rows.Close()

	out := make([]recalledSnippet, 0, limit)
	for rows.Next() {
		var messageID int64
		var excerpt string
		var score float64
		if err := rows.Scan(&messageID, &excerpt, &score); err != nil {
			continue
		}
		excerpt = strings.TrimSpace(excerpt)
		if excerpt == "" || score < contextRecallMinScore {
			continue
		}
		out = append(out, recalledSnippet{
			MessageID: fmt.Sprintf("%d", messageID),
			Score:     score,
			Excerpt:   excerpt,
		})
	}
	return out
}

func ensureThreadMessageEmbeddingsBestEffort(ctx context.Context, client *ent.Client, deps recallDeps, input recallInput) {
	if deps.SQL == nil || deps.Gen == nil {
		return
	}
	ok, err := ai.HasEmbeddingConfig(ctx, client, input.UserID)
	if err != nil || !ok {
		return
	}
	modelCfg, err := ai.ResolveEmbeddingConfig(ctx, client, input.UserID, nil)
	if err != nil {
		return
	}
	exclude := map[int64]struct{}{}
	for _, id := range input.ExcludeMessageIDs {
		exclude[id] = struct{}{}
	}
	rows, err := client.AssistantMessage.Query().
		Where(assistantmessage.ThreadIDEQ(input.ThreadID)).
		Order(ent.Asc(assistantmessage.FieldCreatedAt), ent.Asc(assistantmessage.FieldID)).
		Limit(200).
		All(ctx)
	if err != nil {
		return
	}
	type candidate struct {
		ID      int64
		Excerpt string
	}
	pending := make([]candidate, 0, 32)
	for _, row := range rows {
		if _, skip := exclude[row.ID]; skip {
			continue
		}
		var exists bool
		_ = deps.SQL.QueryRowContext(ctx, `
			select exists(
				select 1 from petrichor_assistant_message_embedding
				where message_id = $1 and embedding is not null
			)
		`, row.ID).Scan(&exists)
		if exists {
			continue
		}
		text := extractPlainFromContentJSON(row.ContentJSON)
		excerpt := sanitizeRecallExcerpt(text)
		if excerpt == "" {
			continue
		}
		pending = append(pending, candidate{ID: row.ID, Excerpt: excerpt})
		if len(pending) >= contextRecallEmbedBatch {
			break
		}
	}
	if len(pending) == 0 {
		return
	}
	texts := make([]string, len(pending))
	for i, c := range pending {
		texts[i] = c.Excerpt
	}
	vecs, err := deps.Gen.Embed(ctx, modelCfg, texts)
	if err != nil || len(vecs) != len(pending) {
		return
	}
	for i, c := range pending {
		if len(vecs[i]) == 0 {
			continue
		}
		literal := vector.FormatPGVector(vecs[i])
		_, _ = deps.SQL.ExecContext(ctx, `
			insert into petrichor_assistant_message_embedding (message_id, thread_id, user_id, excerpt_md, embedding)
			values ($1, $2, $3, $4, $5::vector)
			on conflict (message_id) do update
			set excerpt_md = excluded.excerpt_md,
			    embedding = excluded.embedding
		`, c.ID, input.ThreadID, input.UserID, c.Excerpt, literal)
	}
}

func recallByFTS(ctx context.Context, client *ent.Client, input recallInput, limit int) []recalledSnippet {
	q := client.AssistantMessage.Query().
		Where(assistantmessage.ThreadIDEQ(input.ThreadID)).
		Order(ent.Desc(assistantmessage.FieldCreatedAt), ent.Desc(assistantmessage.FieldID)).
		Limit(80)
	if len(input.ExcludeMessageIDs) > 0 {
		q = q.Where(assistantmessage.IDNotIn(input.ExcludeMessageIDs...))
	}
	rows, err := q.All(ctx)
	if err != nil {
		return nil
	}
	query := strings.ToLower(strings.TrimSpace(input.Query))
	terms := strings.Fields(query)
	if len(terms) == 0 {
		return nil
	}
	type scored struct {
		id      int64
		excerpt string
		score   float64
	}
	scoredRows := make([]scored, 0, limit)
	for _, row := range rows {
		text := extractPlainFromContentJSON(row.ContentJSON)
		if text == "" {
			continue
		}
		lower := strings.ToLower(text)
		hits := 0
		for _, term := range terms {
			if strings.Contains(lower, term) {
				hits++
			}
		}
		if hits == 0 {
			continue
		}
		score := float64(hits) / float64(len(terms))
		if score < contextRecallMinScore {
			continue
		}
		scoredRows = append(scoredRows, scored{
			id:      row.ID,
			excerpt: sanitizeRecallExcerpt(text),
			score:   score,
		})
	}
	for i := 0; i < len(scoredRows); i++ {
		for j := i + 1; j < len(scoredRows); j++ {
			if scoredRows[j].score > scoredRows[i].score {
				scoredRows[i], scoredRows[j] = scoredRows[j], scoredRows[i]
			}
		}
	}
	if len(scoredRows) > limit {
		scoredRows = scoredRows[:limit]
	}
	out := make([]recalledSnippet, 0, len(scoredRows))
	for _, row := range scoredRows {
		out = append(out, recalledSnippet{
			MessageID: fmt.Sprintf("%d", row.id),
			Score:     row.score,
			Excerpt:   row.excerpt,
		})
	}
	return out
}

func extractPlainFromContentJSON(raw *string) string {
	if raw == nil || strings.TrimSpace(*raw) == "" {
		return ""
	}
	var obj map[string]any
	if err := json.Unmarshal([]byte(*raw), &obj); err == nil {
		if text, ok := obj["text"].(string); ok {
			return text
		}
	}
	return strings.TrimSpace(*raw)
}

File: apps/web/src/server/assistant/context-recall.ts
import { asc, eq, sql } from "drizzle-orm"
import { createLogger, toLogError } from "@/lib/logger"
import { embedQuery, embedTexts, hasEmbeddingConfig } from "@/server/ai/embedding"
import { getDb, isSqliteDatabase } from "@/server/db/client"
import { assistantMessages } from "@/server/db/schema"
import { cosineDistance, cosineSimilarity, whereSameDimension } from "@/server/retrieval/vector-space"
import { extractMessagePlainText } from "./message-text"

const log = createLogger("assistant-context-recall")

export const CONTEXT_RECALL_TOP_K = 4
export const CONTEXT_RECALL_MIN_SCORE = 0.25
export const CONTEXT_RECALL_EMBED_BATCH = 16
export const CONTEXT_RECALL_EXCERPT_MAX = 500

export type RecalledSnippet = {
    messageId: string
    score: number
    excerpt: string
}

/** @deprecated 使用 extractMessagePlainText；保留别名兼容 */
export function extractPlainTextForRecall(message: unknown): string {
    return extractMessagePlainText(message)
}

/** 去掉密钥/确认明文等，避免写入召回摘要。 */
export function sanitizeRecallExcerpt(text: string): string {
    let next = text
        .replace(/sk-[a-zA-Z0-9]{10,}/g, "[redacted]")
        .replace(/(api[_-]?key|token|secret|password|cookie)\s*[:=]\s*\S+/gi, "$1=[redacted]")
        .replace(/bearer\s+[a-zA-Z0-9._-]+/gi, "Bearer [redacted]")
        .replace(/executionOutcome[\s\S]{0,200}/gi, "[confirmation omitted]")
    next = next.replace(/\s+/g, " ").trim()
    if (next.length <= CONTEXT_RECALL_EXCERPT_MAX) return next
    return `${next.slice(0, CONTEXT_RECALL_EXCERPT_MAX)}…`
}

export function buildInstructionsWithContextExtras(
    basePrompt: string,
    summaryMd: string | null,
    recalledSnippets?: RecalledSnippet[] | null,
): string {
    const parts = [basePrompt]
    if (summaryMd?.trim()) {
        parts.push(
            "",
            "以下是本对话较早内容的摘要（已折叠细节，仅供连贯理解；最近几轮原文仍在消息中）：",
            summaryMd.trim(),
        )
    }
    if (recalledSnippets?.length) {
        parts.push(
            "",
            "以下是与当前问题相关的较早对话片段（向量召回，供参考，非完整历史）：",
            ...recalledSnippets.map((item, index) =>
                `${index + 1}. (相关度 ${item.score.toFixed(2)}) ${item.excerpt}`),
        )
    }
    return parts.join("\n")
}

export async function recallRelevantHistory(input: {
    userId: number
    threadId: number
    query: string
    excludeMessageIds: number[]
    limit?: number
}): Promise<RecalledSnippet[]> {
    const query = input.query.trim()
    if (!query) return []
    if (isSqliteDatabase()) return []
    try {
        if (!(await hasEmbeddingConfig(input.userId))) return []

        // 热路径只读已有 embedding；缺向量时后台异步补齐，不阻塞本轮召回
        void ensureThreadMessageEmbeddingsBestEffort({
            userId: input.userId,
            threadId: input.threadId,
            excludeMessageIds: input.excludeMessageIds,
        }).catch((error) => {
            log.warn({
                err: toLogError(error),
                userId: input.userId,
                threadId: input.threadId,
            }, "后台补齐消息向量失败")
        })

        const vector = await embedQuery(input.userId, query.slice(0, 4_000))
        if (!vector?.length) return []
        const literal = `[${vector.join(",")}]`
        // 该表没有维度元数据列，按当前向量的实际长度过滤，避免跨维度比较报错
        const dims = vector.length
        const limit = input.limit ?? CONTEXT_RECALL_TOP_K
        const exclude = input.excludeMessageIds.filter((id) => Number.isFinite(id))

        const rows = exclude.length > 0
            ? await getDb().execute(sql`
                select message_id, excerpt_md,
                    (${cosineSimilarity("embedding", literal, dims)})::float8 as score
                from petrichor_assistant_message_embedding
                where thread_id = ${input.threadId}
                  and user_id = ${input.userId}
                  and embedding is not null
                  and ${whereSameDimension("embedding", dims)}
                  and message_id not in (${sql.join(exclude.map((id) => sql`${id}`), sql`, `)})
                order by ${cosineDistance("embedding", literal, dims)}
                limit ${limit}
            `)
            : await getDb().execute(sql`
                select message_id, excerpt_md,
                    (${cosineSimilarity("embedding", literal, dims)})::float8 as score
                from petrichor_assistant_message_embedding
                where thread_id = ${input.threadId}
                  and user_id = ${input.userId}
                  and embedding is not null
                  and ${whereSameDimension("embedding", dims)}
                order by ${cosineDistance("embedding", literal, dims)}
                limit ${limit}
            `)

        const snippets: RecalledSnippet[] = []
        for (const row of rows as unknown as Array<Record<string, unknown>>) {
            const messageId = Number(row.message_id)
            const score = Number(row.score)
            const excerpt = typeof row.excerpt_md === "string" ? row.excerpt_md.trim() : ""
            if (!Number.isFinite(messageId) || !excerpt) continue
            if (!Number.isFinite(score) || score < CONTEXT_RECALL_MIN_SCORE) continue
            snippets.push({
                messageId: String(messageId),
                score,
                excerpt,
            })
        }
        return snippets
    } catch (error) {
        log.warn({
            err: toLogError(error),
            userId: input.userId,
            threadId: input.threadId,
        }, "上下文向量召回失败，跳过本路召回")
        return []
    }
}

async function ensureThreadMessageEmbeddingsBestEffort(input: {
    userId: number
    threadId: number
    excludeMessageIds: number[]
}) {
    const exclude = new Set(input.excludeMessageIds)
    const messages = await getDb()
        .select({
            id: assistantMessages.id,
            role: assistantMessages.role,
            contentJson: assistantMessages.contentJson,
        })
        .from(assistantMessages)
        .where(eq(assistantMessages.threadId, input.threadId))
        .orderBy(asc(assistantMessages.createdAt), asc(assistantMessages.id))
        .limit(200)

    const candidates = messages
        .filter((row) => !exclude.has(row.id))
        .map((row) => {
            let content: unknown = null
            try {
                content = row.contentJson ? JSON.parse(row.contentJson) : null
            } catch {
                content = row.contentJson
            }
            const excerpt = sanitizeRecallExcerpt(extractPlainTextForRecall({
                role: row.role,
                ...(content && typeof content === "object" ? content as object : { content }),
            }))
            return { id: row.id, excerpt }
        })
        .filter((row) => row.excerpt.length >= 8)

    if (candidates.length === 0) return

    const existing = await getDb().execute(sql`
        select message_id
        from petrichor_assistant_message_embedding
        where thread_id = ${input.threadId}
          and user_id = ${input.userId}
    `)
    const embeddedIds = new Set(
        [...(existing as unknown as Array<Record<string, unknown>>)]
            .map((row) => Number(row.message_id))
            .filter((id) => Number.isFinite(id)),
    )

    const pending = candidates
        .filter((row) => !embeddedIds.has(row.id))
        .slice(0, CONTEXT_RECALL_EMBED_BATCH)
    if (pending.length === 0) return

    const vectors = await embedTexts(input.userId, pending.map((row) => row.excerpt))
    for (let i = 0; i < pending.length; i += 1) {
        const vector = vectors[i]
        const row = pending[i]
        if (!vector || !row) continue
        const literal = `[${vector.join(",")}]`
        await getDb().execute(sql`
            insert into petrichor_assistant_message_embedding
                (message_id, thread_id, user_id, excerpt_md, embedding, created_at)
            values
                (${row.id}, ${input.threadId}, ${input.userId}, ${row.excerpt}, ${literal}::vector, now())
            on conflict (message_id) do update set
                excerpt_md = excluded.excerpt_md,
                embedding = excluded.embedding
        `)
    }
}
